status/status.go

```diff
@@ -244,3 +294,153 @@ func printJSON(all []se) {
 	out.Write(b2.Bytes())
 	out.Flush()
 }
+
+type serveResponse struct {
+	Connectivity float64
+	Matrix       map[string]int
+	Self         string
+}
+
+type server struct {
+	response serveResponse
+	mu       sync.Mutex
+
+	si       *network.ServerIdentity
+	list     []*network.ServerIdentity
+	timeout  time.Duration
+	interval time.Duration
+	quit     chan struct{}
+}
+
+func serve(c *cli.Context) error {
+	if c.NArg() != 2 {
+		return errors.New("please give 2 arguments: group.toml private.toml")
+	}
+	ro, err := readGroup(c.Args().First())
+	if err != nil {
+		return errors.New("couldn't read file: " + err.Error())
+	}
+	log.Lvl3(ro)
+	list := ro.List
+	log.Info("List is", list)
+	to, err := time.ParseDuration(c.Parent().String("timeout"))
+	if err != nil {
+		return errors.New("duration parse error: " + err.Error())
+	}
+	coth, err := app.LoadCothority(c.Args().Get(1))
+	if err != nil {
+		return errors.New("error while loading private.toml: " + err.Error())
+	}
+	si, err := coth.GetServerIdentity()
+	if err != nil {
+		return errors.New("private.toml didn't have a serverIdentity: " + err.Error())
+	}
+
+	format := c.String("format")
+	if format != "prometheus" {
+		return errors.New("unsupported format: " + format)
+	}
+
+	endpoint := c.String("endpoint")
+	if len(endpoint) == 0 || endpoint[0] != '/' {
+		return errors.New("invalid endpoint: " + endpoint)
+	}
+
+	port := c.Int("port")
+	interval := c.Int("interval")
+
+	s := &server{
+		interval: time.Duration(interval) * time.Second,
+		timeout:  to,
+		list:     list,
+		si:       si,
+		quit:     make(chan struct{}),
+	}
+
+	http.HandleFunc(endpoint, s.serveHandler)
+	httpServer := &http.Server{Addr: fmt.Sprintf(":%d", port)}
+
+	done := s.probe()
+	go func() {
```

Comment:
There might be a simpler way to run the http server, I've often seen that `ListenAndServe` is called blocking. You can then have something like that, which has slightly less channels:

```go
	signal.Notify(quit, os.Interrupt)

	go func() {
		<-quit
		log.Info("Server is shutting down...")

		ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
		defer cancel()

		if err := server.Shutdown(ctx); err != nil {
			log.Fatalf("Could not gracefully shutdown the server: %v\n", err)
		}
		close(done)
	}()

	log.Infof("Starting HTTP server on %d\n", port)

	err := h.server.ListenAndServe()
	if err != nil && err != http.ErrServerClosed {
		log.Fatalf("Could not listen on %s: %v", h.listenAddr, err)
	}

	<-done
	log.Info("Server stopped")
```

and then you can use `done` in the `probe()`.

Reply:
> and then you can use done in the probe().

wouldn't that lead to a race condition on which goroutine is scheduled to recognise close on done first? Let's say the main goroutine is scheduled first and accepts the channel close. I think it would just log "Server stopped" and exit, thereby leaking the probe goroutine.

Reply:
Once a channel is closed, every subsequent call to it will pass:
https://play.golang.org/p/_N4rtwZ7HVB

Reply:
But the other goroutine might not finish at the same time as main: https://play.golang.org/p/G3TlAARPIcQ and https://play.golang.org/p/sXpLqESWHS2

Reply:
Oh, I see what you mean now. That shouldn't be a problem, as long as the goroutine doesn't have to do some cleanup, you can exit a program with active goroutines: https://stackoverflow.com/questions/25518531/ok-to-exit-program-with-active-goroutine

But this is a detail, I'm perfectly fine with keeping the current implementation.